tests/perf/perf_tests.cc

```diff
@@ -203,6 +204,119 @@ static inline std::ostream& operator<<(std::ostream& os, duration d)
     return os;
 }
 
+/**
+ * A column object encapsulates the logic needed to print one
+ * type of result value, usually as a column (or a json attribute).
+ *
+ * This allows all printers to share a common view of the available
+ * columns.
+ */
+struct column {
+    static constexpr int default_width = 11;
+
+    using print_fn = std::function<void(const column&, std::FILE *file, const result& r)>;
+
+    template <typename F>
+    column(sstring header, int prec, F fn) : header{header}, prec{prec} {
+        using result_t = std::invoke_result_t<F,const result&>;
+        constexpr auto is_integral = std::is_integral_v<result_t>;
+        constexpr auto is_double = std::is_same_v<result_t, double>;
+        constexpr auto is_duration = std::is_same_v<result_t, duration>;
+        if constexpr (is_integral || is_double || is_duration) {
+            static constexpr std::string_view fmt_str = is_double ? "{:>{}.{}f}": "{:>{}}";
+            text_printer = [=](const column& c, std::FILE *file, const result& r) {
+                fmt::print(file, fmt_str, fn(r), c.width, c.prec);
+            };
+        } else {
+            static_assert(false, "unsupported return type");
+        }
+        double_fn = [fn](const result& r) {
+            if constexpr(is_duration) {
+                return fn(r).value;
+            } else {
+                return static_cast<double>(fn(r));
+            }
+        };
+    }
+
+    void print_header(std::FILE *file, const char* str = nullptr) const {
+        fmt::print(file, "{:>{}}", str ? str : header, width);
+    }
+
+    // column header
+    sstring header;
+
+    // width for the column in text output
+    int width = 11;
+
+    // precision in case of double
+    int prec = 3;
+
+    // used by stdout and md formats to print as text
+    print_fn text_printer;
```

Comment:
nit, i'd suggest s/text_printer/print_text/ as we use this member variable as a function, so would be more readable to name it like a verb instead of a noun.